crates/ty_python_semantic/src/types.rs

```diff
@@ -7675,7 +7673,18 @@ impl<'db> TypeVarBoundOrConstraints<'db> {
                 TypeVarBoundOrConstraints::UpperBound(bound.normalized_impl(db, visitor))
             }
             TypeVarBoundOrConstraints::Constraints(constraints) => {
-                TypeVarBoundOrConstraints::Constraints(constraints.normalized_impl(db, visitor))
+                // Constraints are a non-normalized union by design (it's not really a union at
+                // all, we are just using a union to store the types). Normalize the types but not
+                // the containing union.
```

Comment:
And I don't think we _should_ be using a union here, really (though it's a separate issue) -- I'd wager it's the cause of this panic (though I haven't checked that theory)

https://github.com/astral-sh/ruff/blob/0d7ed32494df58f62ed6b48a3491f258ca7fd086/crates/ty_python_semantic/resources/primer/bad.txt#L16-L17

Reply:
the pandas-stubs commit that made us start panicking sure did add a TypeVar with a lot of constraints, soo... https://github.com/pandas-dev/pandas-stubs/commit/bf1221eb7ea0e582c30fe233d1f4f5713fce376b